common/coin_support/solana.c

```diff
@@ -0,0 +1,149 @@
+ * grant to you, the right to Sell the Software.
+ *
+ * For purposes of the foregoing, "Sell" means practicing any or all
+ * of the rights granted to you under the License to provide to third
+ * parties, for a fee or other consideration (including without
+ * limitation fees for hosting or consulting/ support services related
+ * to the Software), a product or service whose value derives, entirely
+ * or substantially, from the functionality of the Software. Any license
+ * notice or attribution required by the License must also include
+ * this Commons Clause License Condition notice.
+ *
+ * Software: All X1Wallet associated files.
+ * License: MIT
+ * Licensor: HODL TECH PTE LTD
+ *
+ ******************************************************************************
+ */
+
+#include "solana.h"
+
+uint16_t get_compact_array_size(const uint8_t *data, uint16_t *size) {
+  uint16_t offset = 0;
+  *size           = 0;
+  while (offset < 3) {
+    *size |= (*(data + offset) & 0x7F) << offset * 7;
+    if ((*(data + offset) & 0x80) == 0)
+      break;
+    offset++;
+  }
+  return offset + 1;
+}
+
+int solana_byte_array_to_unsigned_txn(uint8_t *byte_array, uint16_t byte_array_size, solana_unsigned_txn *utxn) {
+  uint16_t offset = 0;
+
+  // Message header
+  utxn->required_signatures_count                      = *(byte_array + offset++);
+  utxn->read_only_accounts_require_signature_count     = *(byte_array + offset++);
+  utxn->read_only_accounts_not_require_signature_count = *(byte_array + offset++);
+
+  // Account addresses
+  offset += get_compact_array_size(byte_array + offset, &(utxn->account_addresses_count));
+  utxn->account_addresses = byte_array + offset;
+  offset += utxn->account_addresses_count * SOLANA_ACCOUNT_ADDRESS_LENGTH;
+
+  // Blockhash
+  utxn->blockhash = byte_array + offset;
+  offset += SOLANA_BLOCKHASH_LENGTH;
+
+  // Instructions
+  offset += get_compact_array_size(byte_array + offset, &(utxn->instructions_count));
+  if (utxn->instructions_count != 1)
+    return -1;
+
+  utxn->instruction.program_id_index = *(byte_array + offset++);
+
+  uint8_t system_program_id[SOLANA_ACCOUNT_ADDRESS_LENGTH] = {0};  // System instruction address
+  if (memcmp(utxn->account_addresses + utxn->instruction.program_id_index * SOLANA_ACCOUNT_ADDRESS_LENGTH,
+             system_program_id, SOLANA_ACCOUNT_ADDRESS_LENGTH) != 0)
+    return -1;
+
+  offset += get_compact_array_size(byte_array + offset, &(utxn->instruction.account_addresses_index_count));
+  utxn->instruction.account_addresses_index = byte_array + offset;
+  offset += utxn->instruction.account_addresses_index_count;
+  offset += get_compact_array_size(byte_array + offset, &(utxn->instruction.opaque_data_length));
+  utxn->instruction.opaque_data = byte_array + offset;
+  offset += utxn->instruction.opaque_data_length;
+
+  uint32_t instruction_enum = U32_READ_LE_ARRAY(utxn->instruction.opaque_data);
+
+  switch (instruction_enum) {
+    case 2:  // transfer instruction
```

Comment:
in reference to (this comment](https://github.com/Cypherock/x1_wallet_firmware/pull/39#discussion_r985734964), lets use enum for the cases.